gax/src/main/java/com/google/api/gax/rpc/ApiExceptionFactory.java

```diff
@@ -29,15 +29,103 @@
+ *
+ * <p>You can override the default mapping via the {@link ServiceLoader} pattern by declaring an
+ * implementation of {@link ExceptionTransformation}.
  */
 public class ApiExceptionFactory {
-  private ApiExceptionFactory() {}
+
+  private static final ExceptionTransformation DEFAULT_EXCEPTION_TRANSFORMATION =
+      new IdentityExceptionTransformation();
+  private static final ExceptionTransformation transform = loadExceptionTransformation();
 
   public static ApiException createException(
       Throwable cause, StatusCode statusCode, boolean retryable) {
+    ApiException exception = mapStatusToException(cause, statusCode, retryable);
+    exception = transform.transform(exception);
+    return validate(exception, statusCode);
+  }
+
+  public static ApiException createException(
+      String message, Throwable cause, StatusCode statusCode, boolean retryable) {
+    ApiException exception = mapStatusToException(message, cause, statusCode, retryable);
+    exception = transform.transform(exception);
+    return validate(exception, statusCode);
+  }
+
+  /**
+   * Ensures that we will return a valid exception.
+   *
+   * @throws IllegalStateException if the exception is null.
+   */
+  private static ApiException validate(ApiException exception, StatusCode statusCode) {
+    if (exception == null) {
+      throw new IllegalStateException(
+          "Unable to map "
+              + statusCode
+              + " to exception. Possibly a bad ExceptionTransformation service provider?");
+    }
+    return exception;
+  }
+
+  /**
+   * Provides an interface to allow applying an arbitrary transformation to the default exception
+   * mapping.
+   *
+   * <p>This can be useful to map the generic {@link ApiException} implementations to a client
+   * library specific implementation. For instance, to extract server-provided information from the
+   * gRPC trailers and present it to the user.
+   */
+  public interface ExceptionTransformation {
+    /** Transforms a given exception to another exception. */
+    ApiException transform(ApiException exception);
+  }
+
+  /** Default exception transformation which returns the unmodified input. */
+  public static class IdentityExceptionTransformation implements ExceptionTransformation {
+
+    @Override
+    public ApiException transform(ApiException exception) {
+      return exception;
+    }
+  }
+
+  private ApiExceptionFactory() {}
+
+  /**
+   * Loads an implementation of {@link ApiExceptionFactory} via {@link ServiceLoader}. Defers to a
+   * default implementation if no {@link ServiceLoader} is found.
+   *
+   * @return the implementation of {@link ApiExceptionFactory} to use.
+   * @throws IllegalStateException if multiple definitions of {@link ServiceLoader} are found.
+   */
+  private static ExceptionTransformation loadExceptionTransformation() {
+    ServiceLoader<ExceptionTransformation> loader =
+        ServiceLoader.load(ExceptionTransformation.class);
```

Comment:
This looks suspicious. Reflection usage and ingestion of a logic to this class via a very-very indirect way.

Can you please provide more details to justify this change? Specifically:
(1) we don't need to override N different callable types

How many callables there are, what are their names? Are you talking about API-specific callables or the generic ones, defined in gax? If it is the generic ones, there is a very limited  number of those, and overriding those for your API seems like a cleaner approach then trying to ingest custom logic by dynamically loading `ExceptionTransformation` classes. 

Messing with class loaders makes the code less portable but most importantly may lead to unexpected side effects. `ServiceLoader.load(ExceptionTransformation.class)` gives little control on what gets loaded, and looks like, is in the scope of the whole process, not in the scope of the specific client. Is it? If yes, then won't having Ads client and some other google cloud service client in the same process together make the other google cloud service client start using Ads transfomrers?

Reply:
+1. Would you have a link to a sample invocation so we can better look at the surrounding context?

Reply:
Here we need to override all the callables created in [this class](https://github.com/googleads/google-ads-java/blob/master/google-ads/src/main/java/com/google/ads/googleads/lib/GrpcGoogleAdsCallableFactory.java). You can see the implementation of the UnaryCallable callable case [here](https://github.com/googleads/google-ads-java/blob/master/google-ads/src/main/java/com/google/ads/googleads/lib/ExceptionTransformingCallable.java).

Good point about interaction with other (e.g. cloud) clients. We can make the ServiceLoader pattern work here by executing a chain of transformations. Each transformer would only claim exceptions that it's interested in. So in the case of ads, we check the metadata for GoogleAdsFailure instances, and only transform the exceptions that are intended for us.

We can continue overriding the callable creation. However this feels a little heavy handed when we're really only changing the exception transformation. Is there perhaps another solution, e.g. in the generator that would be cleaner?

Reply:
Happy to continue down the callable route. Looking at implementing this, I'd suggest that we push the implementation of [`GrpcCallableFactory`](https://github.com/googleapis/gax-java/blob/master/gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcCallableFactory.java ) into instance methods so that I can override `createBaseUnaryCallable` and `createServerStreamingCallable`. I can leave the existing static methods there as a convenience, and add a method `GrpcCallableFactory.getInstance()` which delegates to a default instance. How does that sound?